src/pymoca/tree.py

```diff
@@ -747,6 +775,494 @@ def _check_import_rules(
+            )
+        # TODO: Check with spec about what to do with extends.class_modification in case of redeclare
+        extend_mod = self._append_modifications(
+            extends.class_modification,
+            modification_environment,
+        )
+        extends_class = self._instantiate_class(
+            extends_class, extend_mod, extends_class.parent, partially
+        )
+        extends_instance = ast.InstanceExtends(
+            visibility=extends.visibility,
+            **extends_class.__dict__,
+        )
+        # TODO: Is there another way to handle `extends.class_modification`? This seems like a hack.
+        # _instantiate_partially removes from extend_mod, but not original modification_environment
+        # Reflect changes in extend_mod back into the passed-in modification env
+        modification_environment.arguments = [
+            arg for arg in modification_environment.arguments if arg in extend_mod.arguments
+        ]
+
+        if extends_instance.type in InstanceTree.BUILTIN_TYPES:
+            if len(parent.extends) > 1:
+                raise ModelicaSemanticError(
+                    "When extending a built-in class (Real, Integer, ...) "
+                    "you cannot extend other classes as well"
+                )
+
+        # TODO: Step 4 Check class lookup before and after extends
+
+        return extends_instance
+
+    def _instantiate_symbol(
+        self,
+        symbol: ast.InstanceSymbol,
+        parent: ast.InstanceClass,
+    ) -> None:
+        """Instantiate given symbol"""
+        # This is step 6 of spec 5.6.1.4 **Recursive instantiation of components**
+        # 6. Components are recursively instantiated
+
+        assert isinstance(symbol, ast.InstanceSymbol)
+        assert isinstance(parent, (ast.InstanceClass, ast.InstanceExtends))
+
+        if symbol.name in InstanceTree.BUILTIN_TYPES:
+            symbol.fully_instantiated = True
+            return
+
+        if not isinstance(symbol.type, ast.InstanceClass):
+            symbol_type = find_name(symbol.type, parent)
+            if symbol_type is None:
+                raise NameLookupError(f"{symbol.type} not found in {parent.full_reference()}")
+            if isinstance(symbol_type, ast.InstanceElement):
+                extend_args = [
+                    arg
+                    for arg in symbol_type.modification_environment.arguments
+                    if not isinstance(arg.value, ast.ShortClassDefinition)
+                ]
+                symbol.modification_environment.arguments += extend_args
+            symbol.type = self._instantiate_class(
+                symbol_type,
+                symbol.modification_environment,
+                symbol_type.parent,
+                False,
+            )
+
+        self._copy_symbol_contents(symbol)
+
+    def _instantiate_partially(
+        self,
+        element: Union[
+            ast.Class,
+            ast.Symbol,
+            ast.InstanceClass,
```

Comment:
Yep, types need to be cleaned up per previous comments.

I like that you are thinking about potential improvements! I like your idea. Something that is bugging me is the close relationship of Class and Symbol. Is there a better way for them too?

Reply:
Well, they used to be rather different because Symbols (or Components) had these things like "nominal"/"value", etc. which an ast.Class did not. In your new way of doing this, with Real/Integer/Boolean/String properly defined as Classes in the root, I don't think ast.Symbol and ast.Class need to be all that different anymore. There might still a difference between "self.classes" and "self.components" (or symbols) I think, but that also might just be semantics?

Reply:
I thought about this and it seems hard to combine them because you can redeclare a component/symbol type. However, the built-in type instances are currently represented in an ugly way (`Class(name="Integer", symbols={"Integer":...}, ...`) and a pain to trudge through in the debugger or in test cases. Perhaps we could merge them after the redeclare step? Just throwing out half-baked ideas.

Reply:
Marking as resolved for the purposes of the review. Created issue #335 to keep it on the radar screen.